graylog2-server/src/main/java/org/graylog/plugins/views/search/rest/ViewsResource.java

```diff
@@ -139,15 +150,67 @@ public ViewDTO get(@ApiParam(name = "id") @PathParam("id") @NotEmpty String id)
     @POST
     @ApiOperation("Create a new view")
     @AuditEvent(type = ViewsAuditEventTypes.VIEW_CREATE)
-    public ViewDTO create(@ApiParam @Valid ViewDTO dto, @Context UserContext userContext) throws ValidationException {
+    public ViewDTO create(@ApiParam @Valid  @NotNull(message = "View is mandatory") ViewDTO dto, @Context UserContext userContext) throws ValidationException {
         if (dto.type().equals(ViewDTO.Type.DASHBOARD)) {
             checkPermission(RestPermissions.DASHBOARDS_CREATE);
         }
+
+        validateIntegrity(dto);
+
         final User user = userContext.getUser();
         final ViewDTO savedDto = dbService.saveWithOwner(dto.toBuilder().owner(user.getName()).build(), user);
         return savedDto;
     }
 
+    private void validateIntegrity(ViewDTO dto) {
+        final Search search = searchDbService.get(dto.searchId()).orElseThrow(() -> new BadRequestException("Search " + dto.searchId() + " doesn't exist"));
+
+        final Set<String> searchQueries = search.queries().stream()
+                .map(Query::id)
+                .collect(Collectors.toSet());
+
+        final Set<String> stateQueries = dto.state().keySet();
+
+        if(!searchQueries.equals(stateQueries)) {
+            throw new BadRequestException("Search queries do not correspond to view/state queries");
+        }
+
+        final Set<String> searchTypes = search.queries().stream()
+                .flatMap(q -> q.searchTypes().stream())
+                .map(SearchType::id)
+                .collect(Collectors.toSet());
+
+
+        final Set<String> stateTypes = dto.state().values().stream()
+                .flatMap(v -> v.widgetMapping().values().stream())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+
+        if(!searchTypes.equals(stateTypes)) {
+            throw new BadRequestException("Search types do not correspond to view/search types");
```

Comment:
If possible, we should also return the ids of search types which are missing from the search here.

Reply:
Done